File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int longest=0;
        Map<Character,Integer> mp=new HashMap<>();
        int last=0;

        for(int i=0;i<s.length();i++){
            char ch=s.charAt(i);
            
            if(mp.containsKey(ch)){
                last=Math.max(last,mp.get(ch)+1);               
            }

            mp.put(ch,i);
            longest=Math.max(longest,(i-last+1));
        }
        return longest;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let mp=new Map();
    let start=0;
    let maxLength=0;

    for(let i=0;i<s.length;i++){
        mp.set(s[i],(mp.get(s[i]) || 0)+1);

        while(mp.get(s[i]) > 1){
            let startchar = s[start];
            mp.set(startchar,(mp.get(startchar)||0)-1);

            if(mp.get(startchar)==0){
                mp.delete(startchar);
            }
            start++;
        }
        maxLength=Math.max(maxLength,i-start+1);
    }
    return maxLength;
};